signac/contrib/job.py

```diff
@@ -818,7 +803,9 @@ def __exit__(self, err_type, err_value, tb):
 
     def __setstate__(self, state):
         self.__dict__.update(state)
-        self.statepoint._parent.jobs.append(self)
+        # Note that we append to a list of jobs rather than replacing to
+        # support transparent id updates between shallow copies of a job.
+        self.statepoint._jobs.append(self)
```

Comment:
Perhaps rename to `_job_references` or something that indicates that all of the elements are actually the _same_ job.

Reply:
They aren't the same job object, they are _shallow copies_ of the same job.

Actually, it's worth noting that your question is relevant to a bug I identified while looking at this code (that is also already present on master). Consider this script:
```
import signac

with signac.TemporaryProject() as tmp_project:
    tmp_project.open_job({'a': 1}).init()
    job1 = next(iter(tmp_project))
    job2 = next(iter(tmp_project))
    print(f"The original ids are {job1.id} and {job2.id}")
    job1.sp.a = 0
    print(f"The new ids are {job1.id} and {job2.id}")
```
Running this script *on master* outputs
```
The original ids are 42b7b4f2921788ea14dac5566e6f06d0 and 42b7b4f2921788ea14dac5566e6f06d0
The new ids are 9bfd29df07674bc4aa960cf661b5acd2 and 42b7b4f2921788ea14dac5566e6f06d0
```
Fixing this link would require some much more sophisticated linkages between all jobs, and even that would only be process-local (I can't see any way to fix this between processes without I/O). I also doubt it's a significant problem in practical uses. Just noting that it is one limitation of assuming that in-memory representations are ever valid in our data model.

Reply:
I don't think that's a bug as long as `job1` and `job2` maintain consistency with their respective id / statepoint.

Reply:
I think it is, at best, unexpected behavior, but I also don't think there's any way to fix this with our current data model without unacceptable amounts of additional file I/O (you'd have to reload from disk every time when checking the id). I'm not proposing any actions here, just pointing out a limitation related to this discussion.